### PR TITLE
Rebase Tags

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

File: go/libraries/doltcore/rebase/rebase.go [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+	// reverse slice: oldest first
+	for i := 0; i < len(history)/2; i++ {
+		opp := len(history) - 1 - i
+		history[i], history[opp] = history[opp], history[i]
+	}
+
+	return history, otherParents, nil
+}
+
+// TODO: optimize this function with a visitedNode set
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/rebase/rebase.go]: I'm leaving the naive implementation for now. Storing the SuperSchema in the commit struct will allow this optimization in the future

File: bats/conflict-detection.bats [hunk 2]
```diff
@@ -16,6 +16,47 @@ teardown() {
...
+    dolt add test
+    dolt commit -m "changes pk=0 c1 t0 11"
+    run dolt merge other
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "Updating" ]] || false
+    [[ "$output" =~ "1 tables changed" ]] || false
+    [[ "$output" =~ "1 rows modified" ]] || false
+
+}
+
```
Comment (andrew-wm-arthur) [inline bats/conflict-detection.bats]: This behavior is necessary for this implementation of rebase. I added this test to make that explicit. 

I envision a different merge behavior where we allow merges into dirty tables, and resolve conflicts if any exists. However, we would need to disambiguate the origin of any changes we want to rebase (column-with-tag creation, column-with-datatype creation, etc.)

Comment (reltuk) [reply bats/conflict-detection.bats]: Yeah, I like this behavior.

File: go/libraries/doltcore/doltdb/doltdb.go [hunk 3]
```diff
@@ -513,6 +583,7 @@ func writeValAndGetRef(ctx context.Context, vrw types.ValueReadWriter, val types
 // ResolveParent returns the n-th ancestor of a given commit (direct parent is index 0). error return value will be
 // non-nil in the case that the commit cannot be resolved, there aren't as many ancestors as requested, or the
 // underlying storage cannot be accessed.
+// TODO: this method is broken
```
Comment (reltuk) [inline go/libraries/doltcore/doltdb/doltdb.go]: Document how? (Non-determinstic parentIdx interactions or something?)

File: go/store/datas/database.go [hunk 4]
```diff
@@ -77,6 +77,8 @@ type Database interface {
 	// of a conflict, Commit returns an 'ErrMergeNeeded' error.
 	Commit(ctx context.Context, ds Dataset, v types.Value, opts CommitOptions) (Dataset, error)
 
+	CommitOrphan(ctx context.Context, v types.Value, opts CommitOptions) (types.Struct, error)
```
Comment (reltuk) [inline go/store/datas/database.go]: `RawCommit`? `NewCommitValue`? `CreateCommitValue`? `CommitWithoutDataset`?

File: go/libraries/doltcore/rebase/rebase.go [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+		if ad.IsDone() {
+			break
+		}
+
+		if err != nil {
+			return types.EmptyMap, err
+		}
+
+		if len(diffs) == 0 {
+			return types.EmptyMap, errors.New("async diff timeout")
```
Comment (reltuk) [inline go/libraries/doltcore/rebase/rebase.go]: Is this really an error?

File: go/libraries/doltcore/rebase/rebase.go [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+		}
+
+		if len(diffs) != 1 {
+			panic("only a single diff requested, multiple returned.  bug in AsyncDiffer")
+		}
+
+		d := diffs[0]
+		if d.KeyValue == nil {
+			panic("lol, wut")
```
Comment (reltuk) [inline go/libraries/doltcore/rebase/rebase.go]: Make this less lolz

File: go/libraries/doltcore/rebase/rebase.go [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+		}
+
+		if d.OldValue != nil && d.NewValue != nil { // update
+			rebasedRowEditor.Set(key, newVal)
+		} else if d.OldValue == nil { // insert
+			rebasedRowEditor.Set(key, newVal)
+		} else if d.NewValue == nil { // delete
+			rebasedRowEditor.Remove(key)
+		} else {
+			panic("bad diff")
```
Comment (reltuk) [inline go/libraries/doltcore/rebase/rebase.go]: `Unexpected commit diff result; all values equal nil.`

**Pull request merged**